File: golang/example/shutdown_test.go
package example

import (
	"fmt"
	"testing"
	"time"

	proxy "github.com/pagespeed-quest/http-playback-proxy/golang"
)

// TestHTTPShutdown tests the HTTP shutdown functionality
func TestHTTPShutdown(t *testing.T) {
	// Test with recording proxy
	t.Run("RecordingProxyHTTPShutdown", func(t *testing.T) {
		tmpDir := t.TempDir()
		controlPort := 19081

		p, err := proxy.StartRecording(proxy.RecordingOptions{
			Port:         0, // Auto-assign port
			InventoryDir: tmpDir,
			ControlPort:  &controlPort,
		})
		if err != nil {
			t.Fatalf("Failed to start recording proxy: %v", err)
		}

		// Give it time to start
		time.Sleep(2 * time.Second)

		if !p.IsRunning() {
			t.Fatal("Proxy should be running")
		}

		fmt.Printf("Recording proxy started on port %d, control port %d\n", p.Port, *p.ControlPort)

		// Stop using HTTP shutdown
		if err := p.Stop(); err != nil {
			t.Fatalf("Failed to stop proxy via HTTP: %v", err)
		}

		// Verify it stopped
		time.Sleep(1 * time.Second)
		if p.IsRunning() {
			t.Fatal("Proxy should have stopped")
		}

		fmt.Println("Recording proxy stopped successfully via HTTP shutdown")
	})

	// Test with playback proxy (needs inventory)
	// This test is commented out as it requires a valid inventory
	/*
		t.Run("PlaybackProxyHTTPShutdown", func(t *testing.T) {
			tmpDir := t.TempDir()
			controlPort := 19082

			// Create a minimal inventory
			inventoryPath := filepath.Join(tmpDir, "index.json")
			inventory := proxy.Inventory{
				Resources: []proxy.Resource{},
			}
			if err := proxy.SaveInventory(inventoryPath, &inventory); err != nil {
				t.Fatalf("Failed to create inventory: %v", err)
			}

			p, err := proxy.StartPlayback(proxy.PlaybackOptions{
				Port:         0, // Auto-assign port
				InventoryDir: tmpDir,
				ControlPort:  &controlPort,
			})
			if err != nil {
				t.Fatalf("Failed to start playback proxy: %v", err)
			}

			// Give it time to start
			time.Sleep(2 * time.Second)

			if !p.IsRunning() {
				t.Fatal("Proxy should be running")
			}

			fmt.Printf("Playback proxy started on port %d, control port %d\n", p.Port, *p.ControlPort)

			// Stop using HTTP shutdown
			if err := p.Stop(); err != nil {
				t.Fatalf("Failed to stop proxy via HTTP: %v", err)
			}

			// Verify it stopped
			time.Sleep(1 * time.Second)
			if p.IsRunning() {
				t.Fatal("Proxy should have stopped")
			}

			fmt.Println("Playback proxy stopped successfully via HTTP shutdown")
		})
	*/
}

File: typescript/src/test/shutdown.test.ts
import { test, describe } from 'node:test';
import assert from 'node:assert';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { startRecording, startPlayback, saveInventory } from '../proxy';
import type { Inventory } from '../types';

/**
 * Helper to create a temporary directory
 */
function createTempDir(prefix: string): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), prefix));
}

/**
 * Helper to sleep for a given number of milliseconds
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

describe('Signal Shutdown', () => {
  test('RecordingProxySignalShutdown', async () => {
    const tmpDir = createTempDir('test-recording-shutdown-');

    try {
      // Start recording proxy with auto-assigned port
      const proxy = await startRecording({
        port: 0,
        inventoryDir: tmpDir,
      });

      // Give it time to fully start
      await sleep(2000);

      // Verify it's running
      assert.strictEqual(proxy.isRunning(), true, 'Proxy should be running');

      console.log(`Recording proxy started on port ${proxy.port}`);

      // Stop using signal-based shutdown (SIGTERM on Unix, CTRL_BREAK on Windows)
      await proxy.stop();

      // Verify it stopped
      await sleep(1000);
      assert.strictEqual(proxy.isRunning(), false, 'Proxy should have stopped');

      console.log('Recording proxy stopped successfully via signal shutdown');
    } finally {
      // Clean up temp directory
      fs.rmSync(tmpDir, { recursive: true, force: true });
    }
  });

  test('PlaybackProxySignalShutdown', async () => {
    const tmpDir = createTempDir('test-playback-shutdown-');

    try {
      // Create a minimal inventory
      const inventoryPath = path.join(tmpDir, 'index.json');
      const inventory: Inventory = {
        resources: [],
      };
      await saveInventory(inventoryPath, inventory);

      // Start playback proxy with auto-assigned port
      const proxy = await startPlayback({
        port: 0,
        inventoryDir: tmpDir,
      });

      // Give it time to fully start
      await sleep(2000);

      // Verify it's running
      assert.strictEqual(proxy.isRunning(), true, 'Proxy should be running');

      console.log(`Playback proxy started on port ${proxy.port}`);

      // Stop using signal-based shutdown (SIGTERM on Unix, CTRL_BREAK on Windows)
      await proxy.stop();

      // Verify it stopped
      await sleep(1000);
      assert.strictEqual(proxy.isRunning(), false, 'Proxy should have stopped');

      console.log('Playback proxy stopped successfully via signal shutdown');
    } finally {
      // Clean up temp directory
      fs.rmSync(tmpDir, { recursive: true, force: true });
    }
  });
});
